### PR TITLE
Docker Regression Tests Github action

### PR DESCRIPTION
**Pull request by agunapal**
## Description

This PR creates a github action to build TorchServe Docker Images and run regression tests inside a docker container

To make this happen,
  - Create a new target for building docker ci image
  - Some of the tests don't run as is inside the container. These need to be fixed. Skipping them for now
  - Run docker regression tests nightly as they take longer to run. Each run takes around 1.5 hours
  
 Fix these Pytests

- [ ] https://github.com/pytorch/serve/issues/2460
- [ ] https://github.com/pytorch/serve/issues/2461 
- [ ] https://github.com/pytorch/serve/issues/2462

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

[2_docker-regression (self-hosted, regression-test-gpu).txt](https://github.com/pytorch/serve/files/11941058/2_docker-regression.self-hosted.regression-test-gpu.txt)
[1_docker-regression (ubuntu-20.04).txt](https://github.com/pytorch/serve/files/11941059/1_docker-regression.ubuntu-20.04.txt)

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: test/pytest/test_example_intel_extension_for_pytorch.py
Comment (msaroufim) [inline test/pytest/test_example_intel_extension_for_pytorch.py:92]: can we give this a shorter name? :)

File: test/pytest/test_torch_compile.py
Comment (msaroufim) [inline test/pytest/test_torch_compile.py:67]: Does this not work? What's the error you get?

Same comment for all of the failing tests curious what's the problem

File: .github/workflows/regression_tests_docker.yml
Comment (msaroufim) [inline .github/workflows/regression_tests_docker.yml:17]: how does the doubly nested list work here?

File: docker/Dockerfile.base
Comment (msaroufim) [inline docker/Dockerfile.base:1]: Docs needd to be updated if we have a new base dockerfile, usually convention is Dockerfile means base

File: docker/build_image.sh
Comment (msaroufim) [inline docker/build_image.sh:153]: How is ci different from dev? Can't we just have those be the same?

File: docker/Dockerfile.ci
Comment (msaroufim) [inline docker/Dockerfile.ci:35]: We might have been doing this previously but this is incorrect, it should be cloning the repo and pulling in a specific branch

Otherwise testing this locally is not meaningful

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: test/pytest/test_torch_compile.py
Comment (agunapal) [reply test/pytest/test_torch_compile.py:67]: The curl commands don't work. All the tests which are using curl commands need to be refactored using requests

File: docker/build_image.sh
Comment (agunapal) [reply docker/build_image.sh:153]: dev will have torchserve starting when container is run. 
ci is running the regression tests script on start

File: docker/Dockerfile.base
Comment (agunapal) [reply docker/Dockerfile.base:1]: Dockerfile is still the main file for building official TorchServe Images. Its not referenced anywhere in the docs . But, let me add a comment that the build script is first building a base image and then building the actual image.

Comment (agunapal) [reply docker/Dockerfile.base:1]: Also, this split is necessary for the docker optimization PR. With that PR, we will also have an additional split to differentiate cpu vs gpu

File: docker/Dockerfile.base
Comment (mreso) [inline docker/Dockerfile.base:77]: I still think there is no real use case for cpu Pytorch in a gpu docker image and we should default to current most supported version.

File: docker/Dockerfile.base
Comment (mreso) [inline docker/Dockerfile.base:99]: What is the intention of this line? Do we remove pip here because the version is creating problems down the line?

File: docker/build_image.sh
Comment (mreso) [reply docker/build_image.sh:153]: hmm, can we at least have the root of both dev and regression images be the same? I.e. having a common dev base and then just add the expose ports and start torchserve separatelty?
Speaking of this. What is the intention of the docker regression test? For me it is to test the functionality of TS inside the docker files we deploy? In that case we should stay as close to the actual images we deploy as possible. Ideally using the actually deployed images as  base and adding onmly stuff needed for the regression test on top to avoid diversion over time.

Comment (mreso) [reply docker/build_image.sh:153]: For me it should loo like
```
Dockerfile -> Docker.run_time
         L -> Docker.dev -> Docker.regression (of use run_time as base for this)
```

File: test/pytest/test_example_scriptable_tokenzier.py
Comment (mreso) [inline test/pytest/test_example_scriptable_tokenzier.py:321]: Why not apply this uniformly to 323 as well?

File: test/pytest/test_torch_compile.py
Comment (mreso) [reply test/pytest/test_torch_compile.py:67]: Is this because curl is not installed or not working?

File: test/regression_tests.py
Comment (mreso) [inline test/regression_tests.py:23]: Whats the intention behind this change?

**Review by mreso — CHANGES_REQUESTED** [review]

File: test/pytest/test_example_intel_extension_for_pytorch.py
Comment (agunapal) [reply test/pytest/test_example_intel_extension_for_pytorch.py]: Done

File: test/pytest/test_torch_compile.py
Comment (agunapal) [reply test/pytest/test_torch_compile.py:67]: curl is installed. curl to localhost doesn't work. Tried various things like --network=host etc. Not sure. We can address this in a different PR

File: docker/Dockerfile.base
Comment (agunapal) [reply docker/Dockerfile.base:1]: Done

File: docker/build_image.sh
Comment (agunapal) [reply docker/build_image.sh:153]: We can. I didn't do it yet since we plan to release next week. If we can push the release by a few weeks, I can test various combinations.

File: docker/Dockerfile.ci
Comment (agunapal) [reply docker/Dockerfile.ci]: Done

File: docker/Dockerfile.base
Comment (agunapal) [reply docker/Dockerfile.base:19]: This is going to change in my next PR to slim

File: docker/Dockerfile.base
Comment (agunapal) [reply docker/Dockerfile.base:77]: Yes, will address this in the PR to reduce CPU/GPU docker images

File: docker/Dockerfile.base
Comment (agunapal) [reply docker/Dockerfile.base:99]: Yes, this was done by another user who refactored our Dockerfile and made sure things work.

File: test/pytest/test_example_scriptable_tokenzier.py
Comment (agunapal) [reply test/pytest/test_example_scriptable_tokenzier.py:321]: Done. Somehow the pytest showed a failure only here

File: test/regression_tests.py
Comment (agunapal) [reply test/regression_tests.py:23]: This is now working inside the docker container. Also, I'm not sure if this is adding any value. We should remove it.

File: docker/Dockerfile.base
Comment (agunapal) [reply docker/Dockerfile.base:31]: Will be resolved in the CPU/GPU optimization PR

File: .github/workflows/regression_tests_docker.yml
Comment (msaroufim) [inline .github/workflows/regression_tests_docker.yml:5]: nit: remove comments

File: .github/workflows/regression_tests_docker.yml
Comment (msaroufim) [inline .github/workflows/regression_tests_docker.yml:18]: why are there 2 github actions? This matrix seems to indicate that you're running both GPU and CPU tests here?

File: .github/workflows/regression_tests_docker.yml
Comment (msaroufim) [inline .github/workflows/regression_tests_docker.yml:35]: Is there now enough memory on non custom runners to build both images? My main worry is that subsequent steps should fail if this step fails

File: docker/Dockerfile
Comment (msaroufim) [inline docker/Dockerfile:88]: do you mean nvidia production image? if so could you add a comment to make this clearer

File: docker/Dockerfile
Comment (msaroufim) [inline docker/Dockerfile:171]: this only works if the branch has been pushed to our repo, IMO this should just use the requirements that are present locally. We want to make sure to support people that want to build our images

File: docker/Dockerfile
Comment (msaroufim) [inline docker/Dockerfile:174]: can you give this a longer name to make sure risk of name conflicts with user code is very low?

File: docker/Dockerfile
Comment (msaroufim) [inline docker/Dockerfile:177]: do we need to do this? This will break BC for people that used to use our dockerfiles. I was thinking we could just docker run in interactive mode for our tests

File: examples/dcgan_fashiongen/create_mar.sh
Comment (msaroufim) [inline examples/dcgan_fashiongen/create_mar.sh:19]: is this some other PR?

File: test/pytest/test_example_intel_extension_for_pytorch.py
Comment (msaroufim) [inline test/pytest/test_example_intel_extension_for_pytorch.py:91]: Could you please create an issue tracker with the skipped tests, these tests are not great but they're all important dependencies ONNX, worker thread affinity and compile

File: test/regression_tests.py
Comment (msaroufim) [inline test/regression_tests.py:24]: are changes to this file from another PR?

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: .github/workflows/regression_tests_docker.yml
Comment (agunapal) [reply .github/workflows/regression_tests_docker.yml:5]: Yes, this is temporary. Will update this once we have the PR approved. We don't have a good way to run this

File: .github/workflows/regression_tests_docker.yml
Comment (agunapal) [reply .github/workflows/regression_tests_docker.yml:35]: These will be executed on 2 different runners. One for CPU and one for GPU

File: docker/Dockerfile
Comment (agunapal) [reply docker/Dockerfile:174]: Done.Changed to `TS_RUN_IN_DOCKER`

File: .github/workflows/regression_tests_docker.yml
Comment (agunapal) [reply .github/workflows/regression_tests_docker.yml:44]: They are on different machines and they are temporary. It should be fine

File: docker/Dockerfile
Comment (agunapal) [reply docker/Dockerfile:88]: No, its TorchServe's production image. We are making use of docker tags to build different images vs having multiple Dockerfile. There is a comment already on L87

File: docker/Dockerfile
Comment (agunapal) [reply docker/Dockerfile:177]: It won't break because there is a target specified here https://github.com/pytorch/serve/blob/issues/fix_regression_tests/docker/build_image.sh#L140

File: examples/dcgan_fashiongen/create_mar.sh
Comment (agunapal) [reply examples/dcgan_fashiongen/create_mar.sh:19]: Initially, I ended up fixing one of the failing tests to make it work in docker. Thought I'll keep it. Should I create a separate PR for this?

File: test/pytest/test_example_intel_extension_for_pytorch.py
Comment (agunapal) [reply test/pytest/test_example_intel_extension_for_pytorch.py:91]: Sure, will create and update the PR with issues. But, we cannot test ONNX because of NVIDIA base image

Comment (agunapal) [reply test/pytest/test_example_intel_extension_for_pytorch.py:91]: I intend to address this when I update Dockerfile.dev, where we can possibly build with nvidia runtime image

File: test/regression_tests.py
Comment (agunapal) [reply test/regression_tests.py:24]: Because we are passing TS repo as a shared volume to Docker container, it was erroring on L23. I can keep it and check for the ENV variable. Should I do this?

File: docker/Dockerfile
Comment (agunapal) [reply docker/Dockerfile:171]: I'm ok either way. This would mean going back to using git clone. Should I make the change?

File: test/pytest/test_example_intel_extension_for_pytorch.py
Comment (agunapal) [reply test/pytest/test_example_intel_extension_for_pytorch.py:91]: Just noticed that ONNX test case is passing in docker :D (It shouldn't)

File: test/regression_tests.py
Comment (agunapal) [reply test/regression_tests.py:24]: or we can use this opportunity to get rid of pygit2?

File: examples/dcgan_fashiongen/create_mar.sh
Comment (msaroufim) [reply examples/dcgan_fashiongen/create_mar.sh:19]: Na it's fine

File: test/regression_tests.py
Comment (msaroufim) [reply test/regression_tests.py:24]: No iz fine

File: docker/Dockerfile
Comment (msaroufim) [reply docker/Dockerfile:171]: Yeah git clone is fine, you can use `--depth` to control size

**Review by msaroufim — APPROVED** [review]
Approved assuming CI is green on those images - for the failing test tracker please create a seperate github issue

File: .github/workflows/regression_tests_docker.yml
Comment (namannandan) [inline .github/workflows/regression_tests_docker.yml:25]: Nit: Why do we need the `|| true` part? Does this command fail in certain conditions and we want to ignore it?

File: .github/workflows/regression_tests_docker.yml
Comment (namannandan) [inline .github/workflows/regression_tests_docker.yml:45]: Nit: Would it be better check for the specific GPU runner attributes/labels instead of absence of `ubuntu`?

**Review by namannandan — APPROVED** [review]

File: .github/workflows/regression_tests_docker.yml
Comment (agunapal) [reply .github/workflows/regression_tests_docker.yml:25]: Yes, there are weird permission issues because of pycache geenrated by docker run. This logic worked for me. Will have to monitor the runs for the next few days

File: .github/workflows/regression_tests_docker.yml
Comment (agunapal) [reply .github/workflows/regression_tests_docker.yml:45]: I tried, the problem is that we need to check for the entire thing: [self.hosted, ...] . I couldnt get it to work , hence went for the negation logic.

**Pull request merged**
